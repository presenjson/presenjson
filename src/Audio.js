import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

class Audio extends PureComponent {
    constructor(...args) {
        super(...args);

        if (!this.props.length && this.props.src && this.props.onLoad) {
            const a = document.createElement('audio');
            a.onloadedmetadata = () => {
                this.props.onLoad(a.duration * 1000);
                a.src = '';
            };
            a.src = this.props.src;
        }
    }

    componentDidMount() {
        const action =
            (this.props.play && this.audio.paused && 'play') || 'pause';
        this.audio[action]();
        this.audio.volume = this.props.volume;
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.play !== this.props.play) {
            const action =
                (nextProps.play && this.audio.paused && 'play') || 'pause';
            this.audio[action]();
        }

        if (nextProps.volume !== this.props.volume) {
            this.audio.volume = nextProps.volume;
        }
    }

    render() {
        const { src, play, ...rest } = this.props;
        return (
            <audio ref={(_) => (this.audio = _)} {...rest}>
                {src && <source src={src} type="audio/mpeg" />}
                {this.props.children}
            </audio>
        );
    }
}

Audio.propTypes = {
    src: PropTypes.string,
    volume: PropTypes.number,
    children: PropTypes.node,
    length: PropTypes.number,
    play: PropTypes.bool,
    onLoad: PropTypes.func
};

Audio.defaultProps = {
    volume: 1
};

export default Audio;
