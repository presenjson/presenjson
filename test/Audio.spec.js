import React from 'react';
import { shallow } from 'enzyme';
import Audio from '../src/Audio';

describe('Audio', () => {
    it('should render', () => {
        expect(shallow(<Audio src="test.mp3" />)).toMatchSnapshot();
    });
});
