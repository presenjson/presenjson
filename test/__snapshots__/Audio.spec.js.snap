// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Audio renders correctly 1`] = `
<audio
  some="prop"
  volume={1}
>
  <source
    src="test.mp3"
    type="audio/mpeg"
  />
</audio>
`;
